src/canvasdrawer.js

```diff
@@ -0,0 +1,1151 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+(function( $ ){
+
+/**
+ * @class CanvasDrawer
+ * @memberof OpenSeadragon
+ * @classdesc Default implementation of CanvasDrawer for an {@link OpenSeadragon.Viewer}.
+ * @param {Object} options - Options for this Drawer.
+ * @param {OpenSeadragon.Viewer} options.viewer - The Viewer that owns this Drawer.
+ * @param {OpenSeadragon.Viewport} options.viewport - Reference to Viewer viewport.
+ * @param {Element} options.element - Parent element.
+ * @param {Number} [options.debugGridColor] - See debugGridColor in {@link OpenSeadragon.Options} for details.
+ */
+
+class CanvasDrawer extends $.DrawerBase{
+    constructor(){
+        super(...arguments);
+
+        // Options
+
+
+
+        /**
+         * 2d drawing context for {@link OpenSeadragon.Drawer#canvas} if it's a &lt;canvas&gt; element, otherwise null.
+         * @member {Object} context
+         * @memberof OpenSeadragon.Drawer#
+         */
+        this.context = this.canvas.getContext( '2d' );
+
+        /**
+         * Sketch canvas used to temporarily draw tiles which cannot be drawn directly
+         * to the main canvas due to opacity. Lazily initialized.
+         */
+        this.sketchCanvas = null;
+        this.sketchContext = null;
+
+        // We force our container to ltr because our drawing math doesn't work in rtl.
+        // This issue only affects our canvas renderer, but we do it always for consistency.
+        // Note that this means overlays you want to be rtl need to be explicitly set to rtl.
+        this.container.dir = 'ltr';
+
+        // Image smoothing for canvas rendering (only if canvas is used).
+        // Canvas default is "true", so this will only be changed if user specified "false".
+        this._imageSmoothingEnabled = true;
+
+    }
+
+    /**
+     * @returns {Boolean} true if canvas is supported by the browser, otherwise false
+     */
+    static isSupported(){
+        return $.supportsCanvas;
+    }
+
+    getType(){
+        return 'canvas';
+    }
+
+    /**
+     * create the HTML element (e.g. canvas, div) that the image will be drawn into
+     * @returns {Element} the canvas to draw into
+     */
+    createDrawingElement(){
```

Comment:
Does this need to be public?

Reply:
I think it'd be ok to make it private. It is actually called by the parent class, not internal to each class itself, but since there is inheritance I think private would be fine - at least, prefixing with an underscore. Not sure how to deal with the doc comment, exactly.

Reply:
This brings up a good topic: Inside OSD we have a different meaning for public vs. private than you would in a normal code base. When I say public, I mean something that's documented as an API that users of the library can/should use. When I say private, I mean something that users of the library shouldn't use. This applies even to functions that one module in the library exposes for use by other modules, so it's not strictly private in the computer science sense.

And of course JavaScript doesn't have private at the language level anyway. To make it "private" we just add `@private` to the doc comment and put an `_` in front of the function name.

Reply:
Hmm... I suppose one angle to think about is whether we intend for developers to be able to write their own `DrawerBase` subclass. If so, maybe it does make sense for any functions that would be overwritten by a subclass to be public. What do you think?

Reply:
> Hmm... I suppose one angle to think about is whether we intend for developers to be able to write their own `DrawerBase` subclass. If so, maybe it does make sense for any functions that would be overwritten by a subclass to be public. What do you think?

I think that's why I made it "public" with no underscore. However, extending `DrawerBase` for a custom drawer is pretty advanced and I think its easy enough to figure out that `_createDrawingElement` needs to be overridden - it'll throw an error message if it isn't done right.

Reply:
Yeah, I think it's fair to assume people won't be making custom drawers. If it becomes a thing we'll cross that bridge when we get to it.